packages/ionio/src/Transaction.ts

```diff
@@ -112,60 +119,82 @@ export class Transaction implements TransactionInterface {
+    const updater = new Updater(this.pset);
+    updater.addInputs([
+      {
+        txid: this.fundingUtxo.txid,
+        txIndex: this.fundingUtxo.vout,
+        sequence,
+        witnessUtxo: this.fundingUtxo.prevout,
+        tapLeafScript: {
           controlBlock,
+          leafVersion,
+          script: Buffer.from(leafToSpend.scriptHex, 'hex'),
         },
-      ],
-    });
-    this.fundingUtxoIndex = this.psbt.data.inputs.length - 1;
+        sighashType: LiquidTransaction.SIGHASH_DEFAULT,
+      },
+    ]);
     // only add unblind data if the prevout of the input is confidential
     if (unblindDataFundingUtxo) {
-      const index = this.psbt.data.inputs.length - 1;
-      this.inputBlindingData.set(index, unblindDataFundingUtxo);
+      this.inputBlinders.push({
+        index: this.fundingUtxoIndex,
+        ...unblindDataFundingUtxo,
+      });
     }
   }
 
   withUtxo(utxo: Output | UnblindedOutput): this {
-    this.psbt.addInput({
-      hash: utxo.txid,
-      index: utxo.vout,
-      witnessUtxo: utxo.prevout,
-    });
+    // instantiate Updater
+    const updater = new Updater(this.pset);
+    // add a new input
+    updater.addInputs([
+      {
+        txid: utxo.txid,
+        txIndex: utxo.vout,
+        witnessUtxo: utxo.prevout,
+        sighashType: LiquidTransaction.SIGHASH_ALL,
+      },
+    ]);
+
+    // get the index of last added input
+    const index = this.pset.inputs.length - 1;
+
     // only add unblind data if the prevout of the input is confidential
     if (utxo && isUnblindedOutput(utxo) && isConfidentialOutput(utxo.prevout)) {
-      const index = this.psbt.data.inputs.length - 1;
-      this.inputBlindingData.set(index, utxo.unblindData);
+      this.inputBlinders.push({
+        index,
+        ...utxo.unblindData,
+      });
     }
+
     return this;
   }
 
   withRecipient(
     recipientAddress: string,
     value: number,
-    assetID: string = this.network.assetHash
+    assetID: string = this.network.assetHash,
+    blinderIndex: number = 0
   ): this {
-    let script = address.toOutputScript(recipientAddress);
-
-    this.psbt.addOutput({
-      script,
-      value: confidential.satoshiToConfidentialValue(value),
-      asset: AssetHash.fromHex(assetID, false).bytes,
-      nonce: Buffer.alloc(0),
-    });
-
-    try {
-      const blindKey = address.fromConfidential(recipientAddress).blindingKey;
-      const index = this.psbt.data.outputs.length - 1;
-      this.outputBlindingPubKeys.set(index, blindKey);
-    } catch (ignore) {}
+    const script = address.toOutputScript(recipientAddress);
+    const blindingKey = address.isConfidential(recipientAddress)
+      ? address.fromConfidential(recipientAddress).blindingKey
+      : undefined;
+
+    // instantiate Updater
+    const updater = new Updater(this.pset);
+
+    // add a new outout.
+    // TODO Check if we should assume the contract owner also shoudl blind all outputs added via withRecipient?
```

Comment:
```suggestion
    // add a new output.
    // TODO Check if we should assume the contract owner also should blind all outputs added via withRecipient?
```

To answer the question, I don't think we need to do such check.
The blinder blinds every output whose blinderIndex matches one of the owned inputs.
Therefore it's up to the user to pass the correct blinder index.

PS. if recipientAddress is unconf there's no need to specify any blinderIndex. Same is valid for OP_RETURN without blindingKey.

Reply:
yes that TODO is old, before I added blinderIndex parameter (I was hardcoding 0, which could be ok for most use cases, but better to expose optional param)